### PR TITLE
Android interview Antonchik

### PR DESCRIPTION
**Pull request by hedinant**

### CODE REVIEW

File: Task.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package interview
+
+abstract class Task(
+    val name: String
+) {
+    var status: TaskStatus = TaskStatus.NotStarted
```
Comment (mosmieri) [inline Task.kt:6]: private field

File: Task.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    var schedule: Long = 0
```
Comment (mosmieri) [inline Task.kt:7]: constructor field

File: Task.kt (hunk 1 above)
Comment (mosmieri) [reply Task.kt:6]: and getter

File: Task.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    var dependsOn: Set<Task> = emptySet()
+
+    abstract fun runTask(remoteHistory: RemoteHistoryService)
+
+    fun finish() {
+        status = TaskStatus.Finished
+        TaskManager.instance.taskFinished(this)
```
Comment (mosmieri) [inline Task.kt:14]: move this fun out of Task class, solid restrictions

File: Task.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    fun reschedule(timestamp: Long) {
+        schedule = timestamp
+        TaskManager.instance.taskFinished(this)
```
Comment (mosmieri) [inline Task.kt:19]: solid restrictions

File: Task.kt (hunk 3 above)
Comment (mosmieri) [reply Task.kt:14]: recreate task by copying

File: Task.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Task) return false
+
+        return schedule == other.schedule && name == other.name
+    }
+
+    override fun hashCode(): Int {
+        var result = name.hashCode()
+        result = 31 * result + status.hashCode()
```
Comment (mosmieri) [inline Task.kt:32]: redundant field

File: TaskManager.kt [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    companion object {
+        lateinit var instance: TaskManager
+    }
+
+    private val queue: LinkedList<Task> = LinkedList()
+    private val running: HashSet<Task> = hashSetOf()
+    private val blocked: HashSet<Task> = hashSetOf()
+    private val taskScope = MainScope()
+
+    init {
```
Comment (mosmieri) [inline TaskManager.kt:22]: getInstance() which returning threadsafe object

File: TaskManager.kt [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        instance = this
+        launchChecker()
```
Comment (mosmieri) [inline TaskManager.kt:24]: move to different method

File: TaskManager.kt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    private val remoteHistory: RemoteHistoryService
+) {
+    companion object {
+        lateinit var instance: TaskManager
+    }
+
+    private val queue: LinkedList<Task> = LinkedList()
+    private val running: HashSet<Task> = hashSetOf()
+    private val blocked: HashSet<Task> = hashSetOf()
+    private val taskScope = MainScope()
```
Comment (mosmieri) [inline TaskManager.kt:20]: Dispatchers.IO + SupervicsorJob

File: TaskManager.kt [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                TaskHistoryRecord(
+                    platform = "android",
+                    record = "task " + task.name + " " + task.schedule + " " +
+                            " started " + System.currentTimeMillis()
+                )
+            )
+        } else {
+            blocked.add(task)
+            task.status = TaskStatus.Blocked
+            uploadHistory(
```
Comment (antonchik-a) [inline TaskManager.kt:42]: code duplication

File: TaskManager.kt [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    private val blocked: HashSet<Task> = hashSetOf()
+    private val taskScope = MainScope()
+
+    init {
+        instance = this
+        launchChecker()
+    }
+
+    fun runTask(task: Task) {
+        if (Collections.disjoint(task.dependsOn, running)) {
```
Comment (mosmieri) [inline TaskManager.kt:28]: move to hashmap in TaskManager

File: TaskManager.kt [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        running.remove(task)
+        uploadHistory(
+            TaskHistoryRecord(
+                platform = "android",
+                record = "task " + task.name + " " + task.schedule + " " +
+                        " finished " + System.currentTimeMillis()
+            )
+        )
+
+        blocked.filter { task1 -> task1.dependsOn.contains(task) }.forEach { task1 ->
```
Comment (antonchik-a) [inline TaskManager.kt:62]: naming

File: TaskManager.kt [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        for (i in 0 until queue.size) {
+            if (task.schedule < queue[i].schedule) {
+                queue.add(i, task)
+                return
+            }
+        }
+    }
+
+    fun isRunning(name: String): Boolean {
+        return running.any { task -> task.name === name }
```
Comment (mosmieri) [inline TaskManager.kt:80]: ==

File: TaskManager.kt [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    private fun launchChecker() {
+        taskScope.launch() {
```
Comment (mosmieri) [inline TaskManager.kt:84]: need optimization for contiuous loop

File: TaskManager.kt [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+            while (true) {
+                while (queue[0].schedule <= System.currentTimeMillis()) {
+                    runTask(queue.removeFirst())
+                }
+            }
+        }
+    }
+
+    private fun uploadHistory(record: TaskHistoryRecord) {
+        val response = remoteHistory.upload(record)
```
Comment (mosmieri) [inline TaskManager.kt:94]: try /catch

**Pull request closed without merging**